poc/src/app/common/shared/components/bar-chart-wrapper/bar-chart-wrapper.component.ts

```diff
@@ -28,8 +41,42 @@ export class BarChartWrapperComponent extends ChartWrapperBaseComponent implemen
 
   override ngOnInit(): void {
     super.ngOnInit()
+    this.chartWrapperPopupsService.init({
+      viewContainerRef: this.viewContainerRef,
+      chart: this.chart,
+      popupsStoreService: this.popupsStoreService,
+      chartId: this.chartId,
+      updatePopup: this.updatePopup,
+      xBarClass: '.c3-event-rect',
+    })
+  }
+
+  constructor(
+    private viewContainerRef: ViewContainerRef,
+    private changeDetectorRef: ChangeDetectorRef,
+    chartWrapperPopupsService: ChartWrapperPopupsService,
+    popupsStoreService: PopupsStoreService
+  ) {
+    super(popupsStoreService, chartWrapperPopupsService)
+    console.time('chart')
+  }
+
+  private calculatePopupX({ bbox, barsWidth, eventRectWidth, popupWidth }): number {
```

Comment:
A similar logic and the same functions are used in other components. Refactoring is needed (e.g., applying the "Strategy" pattern).

Reply:
Fixed